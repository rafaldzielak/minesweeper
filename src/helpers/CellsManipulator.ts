import { Cell, CellState, Coords, Field } from "./Field";

export const incrementNeighbours = (coords: Coords, field: Field): Field => {
  const items = getNeigboursItems(coords);
  for (const item of Object.values(items)) {
    if (checkItemInField(item, field)) {
      const [y, x] = item;
      const cell = field[y][x];
      if (cell < 8) field[y][x] = (cell + 1) as Cell;
    }
  }
  return field;
};

export const getNeigboursItems = ([y, x]: Coords): Record<string, [number, number]> => ({
  top: [y - 1, x],
  topRight: [y - 1, x + 1],
  right: [y, x + 1],
  rightBottom: [y + 1, x + 1],
  bottom: [y + 1, x],
  bottomLeft: [y + 1, x - 1],
  left: [y, x - 1],
  leftTop: [y - 1, x - 1],
});

export const checkItemInField = ([y, x]: Coords, { length }: Field): boolean =>
  y >= 0 && x >= 0 && length - y > 0 && length - x > 0;

export const openCell = (coords: Coords, playerField: Field, gameField: Field): Field => {
  const { empty, bomb, hidden } = CellState;
  const [y, x] = coords;
  const gameCell = gameField[y][x];
  if (gameCell === bomb) throw new Error("Game Over");
  if (gameCell === empty) {
    playerField[y][x] = gameCell;
    const items = getNeigboursItems(coords);
    Object.values(items).forEach((coords) => {
      if (checkItemInField(coords, gameField)) {
        const [y, x] = coords;
        const gameCell = gameField[y][x];
        const playerCell = playerField[y][x];
        if (gameCell === empty && playerCell === hidden)
          playerField = openCell(coords, playerField, gameField);
        if (gameCell < bomb) playerField[y][x] = gameCell;
      }
    });
  }
  playerField[y][x] = gameCell;
  return playerField;
};
