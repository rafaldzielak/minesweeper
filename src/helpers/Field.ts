import { incrementNeighbours } from "./CellsManipulator";

export type Cell = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
export type Field = Cell[][];
export type Coords = [number, number];

export const CellState: Record<string, Cell> = {
  empty: 0,
  bomb: 9,
  hidden: 10,
  flag: 11,
  weakFlag: 12,
};

export const emptyFieldGenerator = (size: number, state: Cell = CellState.empty): Field =>
  new Array(size).fill(null).map(() => new Array(size).fill(state));

export const fieldGenerator = (size: number, probability: number): Field => {
  if (probability > 1 || probability < 0) throw "Probability must be between 0 and 1";
  let unprocessedCells = size * size;
  let restCellsWithBombs = unprocessedCells * probability;
  const result: Field = emptyFieldGenerator(size);

  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      if (restCellsWithBombs / unprocessedCells > Math.random()) {
        result[y][x] = CellState.bomb;
        incrementNeighbours([y, x], result);
        restCellsWithBombs--;
      }
      unprocessedCells--;
    }
  }

  return result;
};

export const generateFieldWithDefaultState = (size: number, state: Cell = CellState.empty): Field =>
  new Array(size).fill(null).map(() => new Array(size).fill(state));
