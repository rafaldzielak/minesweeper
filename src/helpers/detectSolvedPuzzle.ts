import { CellState, Field } from "./Field";

export const detectSolvedPuzzle = (playerField: Field, gameField: Field): [boolean, number] => {
  const { hidden, bomb, flag, weakFlag } = CellState;

  let bombsCounter = 0;
  let flagCounter = 0;
  let detectedBombsCounter = 0;
  let doesFieldHaveHiddenCells = false;

  for (const y of gameField.keys()) {
    for (const x of gameField[y].keys()) {
      const gameCell = gameField[y][x];
      const playerCell = playerField[y][x];
      const isPlayerCellFlag = [flag, weakFlag].includes(playerCell);

      if (playerCell === hidden) doesFieldHaveHiddenCells = true;

      if (isPlayerCellFlag) flagCounter++;

      if (gameCell === bomb) {
        bombsCounter++;
        if (playerCell === flag) detectedBombsCounter++;
      }
    }
  }

  const isPuzzleSolved =
    bombsCounter === detectedBombsCounter && flagCounter === bombsCounter && !doesFieldHaveHiddenCells;

  return [isPuzzleSolved, flagCounter];
};
