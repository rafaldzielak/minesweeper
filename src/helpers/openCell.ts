import { checkItemInField, getNeigboursItems } from "./CellsManipulator";
import { detectSolvedPuzzle } from "./detectSolvedPuzzle";
import { CellState, Coords, Field } from "./Field";

export const openCell = (coords: Coords, playerField: Field, gameField: Field): [Field, boolean, number] => {
  const { empty, bomb, hidden } = CellState;
  const [y, x] = coords;
  const gameCell = gameField[y][x];
  if (gameCell === bomb) throw new Error("Game Over");
  if (gameCell === empty) {
    playerField[y][x] = gameCell;
    const items = getNeigboursItems(coords);
    Object.values(items).forEach((coords) => {
      if (checkItemInField(coords, gameField)) {
        const [y, x] = coords;
        const gameCell = gameField[y][x];
        const playerCell = playerField[y][x];
        if (gameCell === empty && playerCell === hidden)
          [playerField] = openCell(coords, playerField, gameField);
        if (gameCell < bomb) playerField[y][x] = gameCell;
      }
    });
  }
  playerField[y][x] = gameCell;

  const [isSolved, flagCounter] = detectSolvedPuzzle(playerField, gameField);

  return [playerField, isSolved, flagCounter];
};
