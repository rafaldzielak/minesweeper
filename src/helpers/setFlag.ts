import { detectSolvedPuzzle } from "./detectSolvedPuzzle";
import { CellState, Coords, Field } from "./Field";

export const setFlag = (coords: Coords, playerField: Field, gameField: Field): [Field, boolean, number] => {
  const [y, x] = coords;
  const cell = playerField[y][x];

  const { flag, weakFlag, hidden } = CellState;

  switch (cell) {
    case flag:
      playerField[y][x] = weakFlag;
      break;
    case weakFlag:
      playerField[y][x] = hidden;
      break;
    case hidden:
      playerField[y][x] = flag;
      break;
  }

  const [isSolved, flagCounter] = detectSolvedPuzzle(playerField, gameField);

  return [playerField, isSolved, flagCounter];
};
