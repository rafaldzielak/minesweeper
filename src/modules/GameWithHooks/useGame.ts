import { useState, useEffect, useCallback } from "react";

import { Field, CellState, emptyFieldGenerator, fieldGenerator, Coords } from "@/helpers/Field";
import { openCell } from "@/helpers/openCell";

import { LevelNames } from "@/modules/GameSettings";
import { setFlag } from "@/helpers/setFlag";
import { useStatus } from "./useStatus";
import { useSettings } from "./useSettings";

type ReturnType = {
  level: LevelNames;
  isGameOver: boolean;
  isWin: boolean;
  isGameStarted: boolean;
  settings: [number, number];
  playerField: Field;
  gameField: Field;
  time: number;
  flagCounter: number;
  onClick: (coords: Coords) => void;
  onContextMenu: (coords: Coords) => void;
  onChangeLevel: (level: LevelNames) => void;
  onReset: () => void;
};

const useGame = (): ReturnType => {
  const {
    settings: [size, bombs],
    level,
    setLevel,
  } = useSettings();

  const { isGameStarted, isWin, isGameOver, setNewGame, setInProgress, setGameWin, setGameLoose } =
    useStatus();

  const [time, setTime] = useState(0);
  const [flagCounter, setFlagCounter] = useState(0);

  const [playerField, setPlayerField] = useState<Field>(emptyFieldGenerator(size, CellState.hidden));
  const [gameField, setGameField] = useState<Field>(fieldGenerator(size, bombs / (size * size)));

  useEffect(() => {
    if (!isGameStarted) return;
    let interval = setInterval(() => setTime((time) => time + 1), 1000);
    if (isGameOver) clearInterval(interval);
    return () => clearInterval(interval);
  }, [isGameStarted, isGameOver, setTime]);

  const onClick = useCallback(
    (coords: Coords) => {
      if (!isGameStarted) setInProgress();
      try {
        const [newPlayerField, isSolved] = openCell(coords, playerField, gameField);
        if (isSolved) setGameWin();
        setPlayerField([...newPlayerField]);
      } catch (error) {
        setPlayerField([...gameField]);
        setGameLoose();
      }
    },
    [isGameStarted, isGameOver, isWin, level, flagCounter]
  );

  const onContextMenu = useCallback(
    (coords: Coords) => {
      if (!isGameStarted) setInProgress();
      const [newPlayerField, isSolved, newFlagCounter] = setFlag(
        coords,
        playerField,
        gameField,
        flagCounter,
        bombs
      );
      setFlagCounter(newFlagCounter);
      if (isSolved) setGameWin();
      setPlayerField([...newPlayerField]);
    },
    [isGameStarted, isGameOver, isWin, level, flagCounter]
  );

  const resetHandler = ([size, bombs]: [number, number]) => {
    const newGameField = fieldGenerator(size, bombs / (size * size));
    const newPlayerField = emptyFieldGenerator(size, CellState.hidden);
    setTime(0);

    setGameField(newGameField);
    setPlayerField(newPlayerField);
    setNewGame();
  };

  const onChangeLevel = useCallback((level: LevelNames) => {
    const newSettings = setLevel(level);
    resetHandler(newSettings);
  }, []);

  const onReset = useCallback(() => resetHandler([size, bombs]), [size, bombs]);

  return {
    time,
    level,
    isGameOver,
    isWin,
    settings: [size, bombs],
    playerField,
    gameField,
    flagCounter,
    isGameStarted,
    onContextMenu,
    onClick,
    onChangeLevel,
    onReset,
  };
};

export default useGame;
