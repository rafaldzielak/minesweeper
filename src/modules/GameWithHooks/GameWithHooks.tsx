import React, { FC, useCallback } from "react";
import { GameLevels, LevelNames } from "@/modules/GameSettings";
import { Scoreboard } from "@/components/Scoreboard";
import { Grid } from "@/components/Grid/Grid";
import { GameArea, Wrapper, GameOver } from "@/components/Game";
import useGame from "./useGame";

export const GameWithHooks: FC = () => {
  const {
    isGameOver,
    isWin,
    level,
    onChangeLevel,
    onClick,
    onReset,
    onContextMenu,
    playerField,
    settings,
    time,
    flagCounter,
  } = useGame();

  const [, bombs] = settings;

  const onChange = useCallback(
    ({ target: { value: level } }: React.ChangeEvent<HTMLSelectElement>) =>
      onChangeLevel(level as LevelNames),
    []
  );

  return (
    <>
      <GameArea>
        <Scoreboard
          time={time}
          bombs={bombs - flagCounter}
          levels={GameLevels as unknown as string[]}
          defaultLevel={level}
          onChangeLevel={onChange}
          onReset={onReset}
        />
        {isGameOver && <GameOver onClick={onReset} isWin={isWin} />}
        <Grid onClick={onClick} onContextMenu={onContextMenu}>
          {playerField}
        </Grid>
      </GameArea>
    </>
  );
};
