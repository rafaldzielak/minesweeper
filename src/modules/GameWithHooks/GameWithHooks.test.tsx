import React from "react";
import userEvent from "@testing-library/user-event";
import { render, screen, waitFor } from "@testing-library/react";

import { CellState } from "@/helpers/Field";

import { GameWithHooks } from "./GameWithHooks";
import { describe, expect, it, vi } from "vitest";

const { empty: e, hidden: h, bomb: b, flag: f } = CellState;

vi.mock("@/helpers/Field");

describe("GameWithHooks test cases", () => {
  describe("Render behaviour", () => {
    it("Render game field by default", () => {
      const { asFragment } = render(<GameWithHooks />);
      expect(screen.getAllByRole("cell")).toHaveLength(81);
      expect(asFragment()).toMatchSnapshot();
    });

    it("onChange game level handler", async () => {
      render(<GameWithHooks />);
      expect(screen.getAllByRole("cell")).toHaveLength(81);
      userEvent.selectOptions(screen.getByRole("combobox"), "intermediate");
      await waitFor(() => expect(screen.getAllByRole("cell")).toHaveLength(256));
      userEvent.selectOptions(screen.getByRole("combobox"), "expert");
      await waitFor(() => expect(screen.getAllByRole("cell")).toHaveLength(484));
    });
  });

  describe("Open cell test cases", () => {
    it("Open empty cell on the beginner level", async () => {
      render(<GameWithHooks />);
      await userEvent.click(screen.getByTestId("10_0,0"));
      await waitFor(() => expect(screen.getAllByTestId(/^0_/)).toHaveLength(18));
    });
  });
});
