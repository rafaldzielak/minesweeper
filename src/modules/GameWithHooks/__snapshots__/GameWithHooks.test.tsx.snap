// Vitest Snapshot v1

exports[`GameWithHooks test cases > Render behaviour > Render game field by default 1`] = `
<DocumentFragment>
  <div
    class="css-1ct4owg"
  >
    <header
      class="css-dn2347"
    >
      <h1
        class="css-18kvaxm"
      >
        Minesweeper
      </h1>
      <legend
        class="css-17qz6a2"
      >
        <strong>
          Flag: 
        </strong>
        <code
          class="css-12ap3z0"
        >
          <span
            class="css-udfepp"
          >
            right click
          </span>
           + 
          <span
            class="css-9c23k1"
          />
        </code>
      </legend>
    </header>
    <div
      class="css-1bh93n8"
    >
      <div
        class="css-1gexv9m"
      >
        <div
          class="css-d2q1pa"
        >
          0
        </div>
        <div>
          <select
            class="css-14j5jmv"
          >
            <option
              class="css-zcmdxr"
              value="beginner"
            >
              beginner
            </option>
            <option
              class="css-zcmdxr"
              value="intermediate"
            >
              intermediate
            </option>
            <option
              class="css-zcmdxr"
              value="expert"
            >
              expert
            </option>
          </select>
          <button
            class="css-1bva717"
          >
            🙂
          </button>
        </div>
        <div
          class="css-d2q1pa"
        >
          10
        </div>
      </div>
      <div
        class="css-1ye5yxo"
        size="9"
      >
        <div
          class="css-1tt7j02"
          data-testid="10_0,0"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_0,1"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_0,2"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_0,3"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_0,4"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_0,5"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_0,6"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_0,7"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_0,8"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_1,0"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_1,1"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_1,2"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_1,3"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_1,4"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_1,5"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_1,6"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_1,7"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_1,8"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_2,0"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_2,1"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_2,2"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_2,3"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_2,4"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_2,5"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_2,6"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_2,7"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_2,8"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_3,0"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_3,1"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_3,2"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_3,3"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_3,4"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_3,5"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_3,6"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_3,7"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_3,8"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_4,0"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_4,1"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_4,2"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_4,3"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_4,4"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_4,5"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_4,6"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_4,7"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_4,8"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_5,0"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_5,1"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_5,2"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_5,3"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_5,4"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_5,5"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_5,6"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_5,7"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_5,8"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_6,0"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_6,1"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_6,2"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_6,3"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_6,4"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_6,5"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_6,6"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_6,7"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_6,8"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_7,0"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_7,1"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_7,2"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_7,3"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_7,4"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_7,5"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_7,6"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_7,7"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_7,8"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_8,0"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_8,1"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_8,2"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_8,3"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_8,4"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_8,5"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_8,6"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_8,7"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_8,8"
          role="cell"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`GameWithHooks test cases > Render game field by default 1`] = `
<DocumentFragment>
  <div
    class="css-1ct4owg"
  >
    <header
      class="css-dn2347"
    >
      <h1
        class="css-18kvaxm"
      >
        Minesweeper
      </h1>
      <legend
        class="css-17qz6a2"
      >
        <strong>
          Flag: 
        </strong>
        <code
          class="css-12ap3z0"
        >
          <span
            class="css-udfepp"
          >
            right click
          </span>
           + 
          <span
            class="css-9c23k1"
          />
        </code>
      </legend>
    </header>
    <div
      class="css-1bh93n8"
    >
      <div
        class="css-1gexv9m"
      >
        <div
          class="css-d2q1pa"
        >
          0
        </div>
        <div>
          <select
            class="css-14j5jmv"
          >
            <option
              class="css-zcmdxr"
              value="beginner"
            >
              beginner
            </option>
            <option
              class="css-zcmdxr"
              value="intermediate"
            >
              intermediate
            </option>
            <option
              class="css-zcmdxr"
              value="expert"
            >
              expert
            </option>
          </select>
          <button
            class="css-1bva717"
          >
            🙂
          </button>
        </div>
        <div
          class="css-d2q1pa"
        >
          10
        </div>
      </div>
      <div
        class="css-1p4hvro"
      >
        🙁
      </div>
      <div
        class="css-da0epy"
        size="2"
      >
        <div
          class="css-1tt7j02"
          data-testid="10_0,0"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_0,1"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_1,0"
          role="cell"
        />
        <div
          class="css-1tt7j02"
          data-testid="10_1,1"
          role="cell"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
